### PR TITLE
Change hot compact way to level merge

### PR DESCRIPTION
**Pull request by JackieTien97**

### CODE REVIEW

File: server/src/main/java/org/apache/iotdb/db/conf/IoTDBConfig.java [hunk 1]
```diff
@@ -249,24 +250,26 @@
...
-  private boolean enableVm = true;
+  private TsFileManagementStrategy tsFileManagementStrategy = TsFileManagementStrategy.LevelStrategy;
 
   /**
-   * The max vm num of each memtable. When vm num exceeds this, the vm files will merge to one.
+   * Work when tsfile_manage_strategy is level_strategy. The max file num of each level. When file
+   * num exceeds this, the files in one level will merge to one.
    */
-  private int maxVmNum = 5;
+  private int maxFileNumInEachLevel = 5;
```
Comment (samperson1997) [inline server/src/main/java/org/apache/iotdb/db/conf/IoTDBConfig.java:267]: 5 or 10? Be unified before merged : )

File: server/src/main/java/org/apache/iotdb/db/conf/IoTDBConfig.java [hunk 2]
```diff
@@ -249,24 +250,26 @@
...
 
   /**
-   * When vmfiles merge times exceeds this, merge the vmfile to the tsfile.
+   * Work when tsfile_manage_strategy is level_strategy. The max num of level.
    */
-  private int maxMergeChunkNumInTsFile = 25;
+  private int maxLevelNum = 5;
```
Comment (samperson1997) [inline server/src/main/java/org/apache/iotdb/db/conf/IoTDBConfig.java:272]: unified to 5 or 10 to keep same as `iotdb-engine.properties` as well

File: server/src/main/java/org/apache/iotdb/db/conf/IoTDBConfig.java [hunk 3]
```diff
@@ -249,24 +250,26 @@
...
+   * the files to the last level.
    */
   private int mergeChunkPointNumberThreshold = 100000;
 
   /**
-   * Is vm merge enable
+   * TsFile manage strategy, define use which hot compaction strategy
    */
-  private boolean enableVm = true;
+  private TsFileManagementStrategy tsFileManagementStrategy = TsFileManagementStrategy.LevelStrategy;
```
Comment (samperson1997) [inline server/src/main/java/org/apache/iotdb/db/conf/IoTDBConfig.java:261]: Is `tsFileManagementStrategy` a config which could only be modified at the first time starting IoTDB? If so, you may have to add it to `IoTDBConfigCheck` to abandon changing it when restarting.

File: server/src/main/java/org/apache/iotdb/db/engine/tsfilemanagement/HotCompactionMergeTaskPoolManager.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * HotCompactionMergeTaskPoolManager provides a ThreadPool to queue and run all hot compaction
+ * tasks.
+ */
+public class HotCompactionMergeTaskPoolManager implements IService {
+
+  private static final Logger LOGGER = LoggerFactory
```
Comment (samperson1997) [inline server/src/main/java/org/apache/iotdb/db/engine/tsfilemanagement/HotCompactionMergeTaskPoolManager.java:38]:
```suggestion
  private static final Logger logger = LoggerFactory
```

File: server/src/main/java/org/apache/iotdb/db/engine/tsfilemanagement/utils/HotCompactionUtils.java [hunk 5]
```diff
@@ -180,44 +193,52 @@ public static void merge(RestorableTsFileIOWriter writer,
...
           if (newChunkMetadata != null && newChunk != null) {
             writer.writeChunk(newChunk, newChunkMetadata);
+            targetResource.updateStartTime(deviceId, newChunkMetadata.getStartTime());
+            targetResource.updateEndTime(deviceId, newChunkMetadata.getEndTime());
           }
         }
         writer.endChunkGroup();
-        if (vmLogger != null) {
-          vmLogger.logDevice(deviceId, writer.getPos());
+        if (hotCompactionLogger != null) {
```
Comment (samperson1997) [inline server/src/main/java/org/apache/iotdb/db/engine/tsfilemanagement/utils/HotCompactionUtils.java:208]: Could we keep the logger not null before calling this method?

File: server/src/main/java/org/apache/iotdb/db/engine/tsfilemanagement/level/LevelTsFileManagement.java [hunk 6]
```diff
@@ -0,0 +1,540 @@
...
+  private final List<List<TsFileResource>> unSequenceTsFileResources = new CopyOnWriteArrayList<>();
+  private final List<List<TsFileResource>> forkedSequenceTsFileResources = new ArrayList<>();
+  private final List<List<TsFileResource>> forkedUnSequenceTsFileResources = new ArrayList<>();
+
+  public LevelTsFileManagement(String storageGroupName, String storageGroupDir) {
+    super(storageGroupName, storageGroupDir);
+    clear();
+  }
+
+  private void deleteLevelFiles(Collection<TsFileResource> vmMergeTsFiles) {
```
Comment (samperson1997) [inline server/src/main/java/org/apache/iotdb/db/engine/tsfilemanagement/level/LevelTsFileManagement.java:74]: Don't forget to change name of param

File: server/src/main/java/org/apache/iotdb/db/engine/tsfilemanagement/level/LevelTsFileManagement.java [hunk 7]
```diff
@@ -0,0 +1,540 @@
...
+    for (TsFileResource vmMergeTsFile : vmMergeTsFiles) {
+      deleteLevelFile(vmMergeTsFile);
+    }
+    for (int i = 0; i < maxLevelNum; i++) {
+      sequenceTsFileResources.get(i).removeAll(vmMergeTsFiles);
+      unSequenceTsFileResources.get(i).removeAll(vmMergeTsFiles);
+    }
+  }
+
+  private static void deleteLevelFile(TsFileResource seqFile) {
```
Comment (samperson1997) [inline server/src/main/java/org/apache/iotdb/db/engine/tsfilemanagement/level/LevelTsFileManagement.java:85]: Why static?

File: server/src/main/java/org/apache/iotdb/db/engine/tsfilemanagement/level/LevelTsFileManagement.java [hunk 8]
```diff
@@ -0,0 +1,540 @@
...
+    seqFile.writeLock();
+    try {
+      ChunkMetadataCache.getInstance().remove(seqFile);
+      FileReaderManager.getInstance().closeFileAndRemoveReader(seqFile.getTsFilePath());
+      seqFile.setDeleted(true);
+      if (seqFile.getTsFile().exists()) {
+        Files.delete(seqFile.getTsFile().toPath());
+      }
+    } catch (Exception e) {
```
Comment (samperson1997) [inline server/src/main/java/org/apache/iotdb/db/engine/tsfilemanagement/level/LevelTsFileManagement.java:94]:
```suggestion
    } catch (IOException e) {
```

File: server/src/main/java/org/apache/iotdb/db/engine/tsfilemanagement/utils/HotCompactionUtils.java (hunk 5 above)
Comment (zhanglingzhe0820) [reply server/src/main/java/org/apache/iotdb/db/engine/tsfilemanagement/utils/HotCompactionUtils.java:208]: just confirm that the logger is not null

**Review by qiaojialin — APPROVED** [review]

**Pull request merged**